include/deal.II/non_matching/quadrature_generator.h

```diff
@@ -0,0 +1,211 @@
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_non_matching_quadrature_generator
+#define dealii_non_matching_quadrature_generator
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/point.h>
+
+#include <type_traits>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace NonMatching
+{
+  namespace internal
+  {
+    namespace QuadratureGeneratorImplementation
+    {
+      /**
+       * Describes a dim-dimensional hypercube aligned with the coordinate
+       * axes. That is, a region
+       *
+       * [x_i - L/2, x_i + L/2]^dim
+       *
+       * where L is the side length and (x_0, ..., x_{dim-1}) are the
+       * coordinates of the center of the hypercube.
+       *
+       * The purpose of this class is to help us keep track of where we are on
+       * the reference cell when generating immersed quadrature rules. There are
+       * two main things we want to be able to do.
+       *
+       * 1) Split a hypercube into its 2^dim children.
+       * 2) Take a cross section orthogonal to a given direction and get this as
+       *    a Hypercube<dim-1>.
+       *
+       * These are needed because the algorithm is allowed to recurse both over
+       * children and over dimensions.
+       *
+       * In 3D, the 2 coordinates of the cross section of HyperCube<3> can be
+       * ordered in 2 ways. That is, if we take the cross section orthogonal to
+       * the y direction we could either order a 3D-coordinate into a
+       * 2D-coordinate as (x,z) or as (z,x). This class uses the second
+       * convention, corresponding to the coordinates being ordered cyclicly
+       * x -> y -> z -> x -> ... To be precise, if we take a cross section:
+       * ---------------------------------------------------------------
+       * | Orthogonal to  |   Cross section coordinates are ordered as |
+       * |-------------------------------------------------------------|
+       * |      x         |                 (y, z)                     |
+       * |      y         |                 (z, x)                     |
+       * |      z         |                 (x, y)                     |
+       * ---------------------------------------------------------------
+       * This is according to the convention set by the function
+       * coordinate_to_one_dim_higher.
+       */
+      template <int dim>
+      class Hypercube
+      {
+      public:
+        /**
+         * Type defining a cross section of a hypercube. The purpose of the
+         * conditional is to avoid that a 1-dimensional hypercube has a cross
+         * section. In 1D, we let the cross section be void.
+         */
+        using CrossSection =
+          typename std::conditional<(1 < dim), Hypercube<dim - 1>, void>::type;
```

Comment:
Its more typical in the library that we have a 0-dim base case instead of doing something like that. Would you mind changing this to do that?

Reply:
Sure, but I don't understand what I need to do to terminate the recursion here. :thinking: I looked at 
`Quadratrue<dim>::SubQuadrature`, 
which is essentially equivalent to this case:
`using SubQuadrature = Quadrature<dim - 1>;`
I would expect that `Quadrature<0>` creates a `Quadrature<-1>`, which wouldn't compile since it requires a `Point<-1>`. But clearly this doesn't happen. Why?

Reply:
This compiles:
```
template <int dim>
class X
{
public:
  using sub = X<dim-1>;

  sub get();
};


int main()
{
  X<2> x;
  x.get().get().get();
}
```
This is because calling `get()` instantiates `X<dim-1>` as a class, but unless you create an object of type `sub`, there is no need for the compiler to create instantiate `X<dim-2>`. In other words, the call to `.get().get().get()` does create `X<0>`, but just because that type has a `using/typedef` for `X<-1>` does not imply that `X<-1>` is actually instantiated.